src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -35,6 +35,494 @@ extern "C" {
 using namespace gpdxl;
 using namespace gpmd;
 
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::GroupingListContainsPrimaryKey
+//
+//	@doc:
+//		Checks if list grouping_list contains all columns from conkeys of
+//		Primary Key in relation conrelid. grouping_list is treated as flat
+//		list of SortGroupClauses, other nodes are ignored.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
+											   List *grouping_list,
+											   Bitmapset *conkeys, Oid conrelid)
+{
+	BOOL result = false;
+	Bitmapset *grouping_set = NULL;
+	ListCell *lgc;
+
+	// Check if conkeys set is a subset of grouping columns set.
+	ForEach(lgc, grouping_list)
+	{
+		Node *grpcl = (Node *) lfirst(lgc);
+
+		if (NULL == grpcl || !IsA(grpcl, SortGroupClause))
+		{
+			continue;
+		}
+
+		SortGroupClause *sgc = (SortGroupClause *) grpcl;
+		Node *expr = gpdb::GetSortGroupClauseExpr(sgc, query->targetList);
+
+		GPOS_ASSERT(IsA(expr, Var));
+
+		Var *var = (Var *) expr;
+
+		RangeTblEntry *rte =
+			(RangeTblEntry *) gpdb::ListNth(query->rtable, var->varno - 1);
+
+		GPOS_ASSERT(NULL != rte);
+
+		if (0 == var->varlevelsup && rte->relid == conrelid)
+		{
+			grouping_set = gpdb::BmsAddMember(grouping_set, var->varattno);
+		}
+	}
+
+	result = gpdb::BmsIsSubset(conkeys, grouping_set);
+
+	gpdb::BmsFree(grouping_set);
+
+	return result;
+}
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::AddMissingGroupClauseMutator
+//
+//	@doc:
+//		Traverses the groupClause, adds new SortGroupClause (defined by m_gc
+//		field in context) to all grouping lists where m_gc has functional
+//		dependency.
+//---------------------------------------------------------------------------
+Node *
+CQueryMutators::AddMissingGroupClauseMutator(
+	Node *node, SContextFixGroupDependentTargets *context)
+{
+	if (NULL == node)
+	{
+		return NULL;
+	}
+
+	if (IsA(node, SortGroupClause))
+	{
+		if (context->m_parent_is_grouping_clause)
+		{
+			// In case the SortGroupClause is direct element of groupsets list
+			// of GroupingClause, we should treat it in a special way - consider
+			// it like a 1-element list.
+			List *new_list = ListMake1((Node *) gpdb::CopyObject(node));
+			if (GroupingListContainsPrimaryKey(context->m_query, new_list,
```

Comment:
How about such optimization
```diff
diff --git a/src/backend/gpopt/translate/CQueryMutators.cpp b/src/backend/gpopt/translate/CQueryMutators.cpp
index 0252500fbe6..8de3c39c139 100644
--- a/src/backend/gpopt/translate/CQueryMutators.cpp
+++ b/src/backend/gpopt/translate/CQueryMutators.cpp
@@ -36,6 +36,22 @@ using namespace gpdxl;
 using namespace gpmd;
 
 
+Var *
+SortGroupClauseContainsPrimaryKey(Query *query, SortGroupClause *sgc, Oid conrelid)
+{
+	Node *expr = gpdb::GetSortGroupClauseExpr(sgc, query->targetList);
+
+	GPOS_ASSERT(IsA(expr, Var));
+
+	Var *var = (Var *) expr;
+
+	RangeTblEntry *rte =
+		(RangeTblEntry *) gpdb::ListNth(query->rtable, var->varno - 1);
+
+	GPOS_ASSERT(NULL != rte);
+
+	return (0 == var->varlevelsup && rte->relid == conrelid ? var : NULL);
+}
 //---------------------------------------------------------------------------
 //	@function:
 //		CQueryMutators::GroupingListContainsPrimaryKey
@@ -64,19 +80,9 @@ CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
 			continue;
 		}
 
-		SortGroupClause *sgc = (SortGroupClause *) grpcl;
-		Node *expr = gpdb::GetSortGroupClauseExpr(sgc, query->targetList);
-
-		GPOS_ASSERT(IsA(expr, Var));
-
-		Var *var = (Var *) expr;
-
-		RangeTblEntry *rte =
-			(RangeTblEntry *) gpdb::ListNth(query->rtable, var->varno - 1);
-
-		GPOS_ASSERT(NULL != rte);
+		Var *var = SortGroupClauseContainsPrimaryKey(query, (SortGroupClause *)grpcl, conrelid);
 
-		if (0 == var->varlevelsup && rte->relid == conrelid)
+		if (NULL != var)
 		{
 			grouping_set = gpdb::BmsAddMember(grouping_set, var->varattno);
 		}
@@ -114,16 +120,11 @@ CQueryMutators::AddMissingGroupClauseMutator(
 			// In case the SortGroupClause is direct element of groupsets list
 			// of GroupingClause, we should treat it in a special way - consider
 			// it like a 1-element list.
-			List *new_list = ListMake1((Node *) gpdb::CopyObject(node));
-			if (GroupingListContainsPrimaryKey(context->m_query, new_list,
-											   context->m_conkeys,
-											   context->m_conrelid))
+			Var *var = SortGroupClauseContainsPrimaryKey(context->m_query, (SortGroupClause *)node, context->m_conrelid);
+			if (NULL != var && 1 == bms_num_members(context->m_conkeys) && bms_is_member(var->varattno, context->m_conkeys))
 			{
-				new_list = gpdb::LAppend(
-					new_list, gpdb::CopyObject((Node *) context->m_gc));
-				return (Node *) new_list;
+				return (Node *) ListMake2((Node *) gpdb::CopyObject(node), gpdb::CopyObject((Node *) context->m_gc));
 			}
-			gpdb::ListFreeDeep(new_list);
 		}
 		return (Node *) gpdb::CopyObject(node);
 	}
```

Reply:
or return to lists
```diff
diff --git a/src/backend/catalog/pg_constraint.c b/src/backend/catalog/pg_constraint.c
index 527509844c9..d474cd23388 100644
--- a/src/backend/catalog/pg_constraint.c
+++ b/src/backend/catalog/pg_constraint.c
@@ -839,12 +839,12 @@ get_constraint_relation_oids(Oid constraint_oid, Oid *conrelid, Oid *confrelid)
 /*
  * get_constraint_relation_columns
  *		Find the columns of the relations to which a constraint refers.
- *		Returns the Bitmapset of found columns.
+ *		Returns the list of found columns.
  */
-Bitmapset *
+List *
 get_constraint_relation_columns(Oid constraint_oid)
 {
-	Bitmapset  *conkeys_set = NULL;
+	List	   *conkeys_list = NIL;
 	HeapTuple	tp = SearchSysCache1(CONSTROID, ObjectIdGetDatum(constraint_oid));
 
 	if (HeapTupleIsValid(tp))
@@ -871,7 +871,7 @@ get_constraint_relation_columns(Oid constraint_oid)
 			{
 				AttrNumber	attnum = DatumGetInt16(conkeys[i]);
 
-				conkeys_set = bms_add_member(conkeys_set, attnum);
+				conkeys_list = lappend_int(conkeys_list, attnum);
 			}
 		}
 
@@ -879,7 +879,7 @@ get_constraint_relation_columns(Oid constraint_oid)
 		ReleaseSysCache(tp);
 	}
 
-	return conkeys_set;
+	return conkeys_list;
 }
 
 /*
diff --git a/src/backend/gpopt/gpdbwrappers.cpp b/src/backend/gpopt/gpdbwrappers.cpp
index 4fc0a7cf35e..394bf5d2cce 100644
--- a/src/backend/gpopt/gpdbwrappers.cpp
+++ b/src/backend/gpopt/gpdbwrappers.cpp
@@ -2845,7 +2845,7 @@ gpdb::GetConstraintRelationOids(Oid constraint_oid, Oid *conrelid,
 	GP_WRAP_END;
 }
 
-Bitmapset *
+List *
 gpdb::GetConstraintRelationColumns(Oid constraint_oid)
 {
 	GP_WRAP_START;
@@ -2853,28 +2853,7 @@ gpdb::GetConstraintRelationColumns(Oid constraint_oid)
 		return get_constraint_relation_columns(constraint_oid);
 	}
 	GP_WRAP_END;
-	return NULL;
-}
-
-void
-gpdb::BmsFree(Bitmapset *a)
-{
-	GP_WRAP_START;
-	{
-		bms_free(a);
-	}
-	GP_WRAP_END;
-}
-
-bool
-gpdb::BmsIsSubset(const Bitmapset *a, const Bitmapset *b)
-{
-	GP_WRAP_START;
-	{
-		return bms_is_subset(a, b);
-	}
-	GP_WRAP_END;
-	return false;
+	return NIL;
 }
 
 // EOF
diff --git a/src/backend/gpopt/translate/CQueryMutators.cpp b/src/backend/gpopt/translate/CQueryMutators.cpp
index 0252500fbe6..5a1985fc12a 100644
--- a/src/backend/gpopt/translate/CQueryMutators.cpp
+++ b/src/backend/gpopt/translate/CQueryMutators.cpp
@@ -36,6 +36,22 @@ using namespace gpdxl;
 using namespace gpmd;
 
 
+AttrNumber
+SortGroupClauseContainsPrimaryKey(Query *query, SortGroupClause *sgc, Oid conrelid)
+{
+	Node *expr = gpdb::GetSortGroupClauseExpr(sgc, query->targetList);
+
+	GPOS_ASSERT(IsA(expr, Var));
+
+	Var *var = (Var *) expr;
+
+	RangeTblEntry *rte =
+		(RangeTblEntry *) gpdb::ListNth(query->rtable, var->varno - 1);
+
+	GPOS_ASSERT(NULL != rte);
+
+	return (var->varlevelsup == 0 && rte->relid == conrelid ? var->varattno : InvalidAttrNumber);
+}
 //---------------------------------------------------------------------------
 //	@function:
 //		CQueryMutators::GroupingListContainsPrimaryKey
@@ -48,13 +64,12 @@ using namespace gpmd;
 BOOL
 CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
 											   List *grouping_list,
-											   Bitmapset *conkeys, Oid conrelid)
+											   List *conkeys, Oid conrelid)
 {
-	BOOL result = false;
-	Bitmapset *grouping_set = NULL;
 	ListCell *lgc;
+	ULONG found_col_cnt = 0;
 
-	// Check if conkeys set is a subset of grouping columns set.
+	// search for conkey in grouping list
 	ForEach(lgc, grouping_list)
 	{
 		Node *grpcl = (Node *) lfirst(lgc);
@@ -64,29 +79,14 @@ CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
 			continue;
 		}
 
-		SortGroupClause *sgc = (SortGroupClause *) grpcl;
-		Node *expr = gpdb::GetSortGroupClauseExpr(sgc, query->targetList);
-
-		GPOS_ASSERT(IsA(expr, Var));
-
-		Var *var = (Var *) expr;
-
-		RangeTblEntry *rte =
-			(RangeTblEntry *) gpdb::ListNth(query->rtable, var->varno - 1);
-
-		GPOS_ASSERT(NULL != rte);
-
-		if (0 == var->varlevelsup && rte->relid == conrelid)
+		if (list_member_int(conkeys, SortGroupClauseContainsPrimaryKey(query, (SortGroupClause *) grpcl, conrelid)))
 		{
-			grouping_set = gpdb::BmsAddMember(grouping_set, var->varattno);
+			found_col_cnt++;
 		}
 	}
 
-	result = gpdb::BmsIsSubset(conkeys, grouping_set);
-
-	gpdb::BmsFree(grouping_set);
-
-	return result;
+	// check that all conkeys were found
+	return (gpdb::ListLength(conkeys) == found_col_cnt);
 }
 
 //---------------------------------------------------------------------------
@@ -114,16 +114,10 @@ CQueryMutators::AddMissingGroupClauseMutator(
 			// In case the SortGroupClause is direct element of groupsets list
 			// of GroupingClause, we should treat it in a special way - consider
 			// it like a 1-element list.
-			List *new_list = ListMake1((Node *) gpdb::CopyObject(node));
-			if (GroupingListContainsPrimaryKey(context->m_query, new_list,
-											   context->m_conkeys,
-											   context->m_conrelid))
+			if (1 == list_length(context->m_conkeys) && list_member_int(context->m_conkeys, SortGroupClauseContainsPrimaryKey(context->m_query, (SortGroupClause *) node, context->m_conrelid)))
 			{
-				new_list = gpdb::LAppend(
-					new_list, gpdb::CopyObject((Node *) context->m_gc));
-				return (Node *) new_list;
+				return (Node *) ListMake2((Node *) gpdb::CopyObject(node), gpdb::CopyObject((Node *) context->m_gc));
 			}
-			gpdb::ListFreeDeep(new_list);
 		}
 		return (Node *) gpdb::CopyObject(node);
 	}
@@ -475,7 +469,7 @@ CQueryMutators::FixGroupDependentTargets(Query *query)
 			gpdb::ListFreeDeep(old_group_clause);
 			gpdb::GPDBFree(gc);
 		}
-		gpdb::BmsFree(ctx_fix_dependent_targets.m_conkeys);
+		gpdb::ListFree(ctx_fix_dependent_targets.m_conkeys);
 	}
 	// remove constraintDeps
 	gpdb::ListFree(query->constraintDeps);
diff --git a/src/include/catalog/pg_constraint.h b/src/include/catalog/pg_constraint.h
index 3a9304116eb..bfcdcad8fb5 100644
--- a/src/include/catalog/pg_constraint.h
+++ b/src/include/catalog/pg_constraint.h
@@ -264,7 +264,7 @@ extern void ConstraintSetParentConstraint(Oid childConstrId,
 extern bool ConstraintGetPrimaryKeyOf(Oid relid, AttrNumber attno, 
 					Oid *pkrelid, AttrNumber *pkattno);
 extern void get_constraint_relation_oids(Oid constraint_oid, Oid *conrelid, Oid *confrelid);
-extern Bitmapset *get_constraint_relation_columns(Oid constraint_oid);
+extern List *get_constraint_relation_columns(Oid constraint_oid);
 extern Oid	get_relation_constraint_oid(Oid relid, const char *conname, bool missing_ok);
 extern Oid	get_domain_constraint_oid(Oid typid, const char *conname, bool missing_ok);
 extern Oid	get_relation_idx_constraint_oid(Oid relationId, Oid indexId);
diff --git a/src/include/gpopt/gpdbwrappers.h b/src/include/gpopt/gpdbwrappers.h
index 355c852f8f1..d1b582f6854 100644
--- a/src/include/gpopt/gpdbwrappers.h
+++ b/src/include/gpopt/gpdbwrappers.h
@@ -740,13 +740,7 @@ void GetConstraintRelationOids(Oid constraint_oid, Oid *conrelid,
 							   Oid *confrelid);
 
 // find the columns of the relations to which a constraint refers
-Bitmapset *GetConstraintRelationColumns(Oid constraint_oid);
-
-// free a Bitmapset
-void BmsFree(Bitmapset *a);
-
-// check if A is a subset of B
-bool BmsIsSubset(const Bitmapset *a, const Bitmapset *b);
+List *GetConstraintRelationColumns(Oid constraint_oid);
 
 }  //namespace gpdb
 
diff --git a/src/include/gpopt/translate/CQueryMutators.h b/src/include/gpopt/translate/CQueryMutators.h
index a6b832c961f..09e017b2228 100644
--- a/src/include/gpopt/translate/CQueryMutators.h
+++ b/src/include/gpopt/translate/CQueryMutators.h
@@ -188,8 +188,8 @@ class CQueryMutators
 		Query *m_query;
 		// primary key relation id
 		Oid m_conrelid;
-		// set of columns that define primary key in relation with m_conrelid
-		Bitmapset *m_conkeys;
+		// list of columns that define primary key in relation with m_conrelid
+		List *m_conkeys;
 		// the current query level
 		ULONG m_current_query_level;
 		// new SortGroupClause to be added into groupClause
@@ -205,7 +205,7 @@ class CQueryMutators
 		SContextFixGroupDependentTargets(Query *query)
 			: m_query(query),
 			  m_conrelid(0),
-			  m_conkeys(NULL),
+			  m_conkeys(0),
 			  m_current_query_level(0),
 			  m_gc(NULL),
 			  m_parent_is_grouping_clause(false),
@@ -217,8 +217,7 @@ class CQueryMutators
 private:
 	static BOOL GroupingListContainsPrimaryKey(Query *query,
 											   List *grouping_list,
-											   Bitmapset *conkeys,
-											   Oid conrelid);
+											   List *conkeys, Oid conrelid);
 
 	static BOOL GetGroupUniqueTleReferencesWalker(
 		Node *node, SContexGetGroupUniqueTleWalker *context);
```

Reply:
> How about such optimization

Did you check this code? Does it handle properly cases when there is > 1 columns in primary key?

Reply:
tests passed

Reply:
(I prefer return lists by last pointed diff)

Reply:
> And can it handle duplicate grouping columns like in below query `explain select b from test_table1 group by grouping sets((a,a), ());` ?

hmm... no, may be add this case to tests?

Reply:
> hmm... no, may be add this case to tests?

Yes, I also think it would be good to add to the tests.

Reply:
fixed
```diff
diff --git a/src/backend/catalog/pg_constraint.c b/src/backend/catalog/pg_constraint.c
index 527509844c9..3451151b142 100644
--- a/src/backend/catalog/pg_constraint.c
+++ b/src/backend/catalog/pg_constraint.c
@@ -839,12 +839,12 @@ get_constraint_relation_oids(Oid constraint_oid, Oid *conrelid, Oid *confrelid)
 /*
  * get_constraint_relation_columns
  *		Find the columns of the relations to which a constraint refers.
- *		Returns the Bitmapset of found columns.
+ *		Returns the list of found columns.
  */
-Bitmapset *
+List *
 get_constraint_relation_columns(Oid constraint_oid)
 {
-	Bitmapset  *conkeys_set = NULL;
+	List	   *conkeys_list = NIL;
 	HeapTuple	tp = SearchSysCache1(CONSTROID, ObjectIdGetDatum(constraint_oid));
 
 	if (HeapTupleIsValid(tp))
@@ -871,7 +871,7 @@ get_constraint_relation_columns(Oid constraint_oid)
 			{
 				AttrNumber	attnum = DatumGetInt16(conkeys[i]);
 
-				conkeys_set = bms_add_member(conkeys_set, attnum);
+				conkeys_list = list_append_unique_int(conkeys_list, attnum);
 			}
 		}
 
@@ -879,7 +879,7 @@ get_constraint_relation_columns(Oid constraint_oid)
 		ReleaseSysCache(tp);
 	}
 
-	return conkeys_set;
+	return conkeys_list;
 }
 
 /*
diff --git a/src/backend/gpopt/gpdbwrappers.cpp b/src/backend/gpopt/gpdbwrappers.cpp
index 4fc0a7cf35e..394bf5d2cce 100644
--- a/src/backend/gpopt/gpdbwrappers.cpp
+++ b/src/backend/gpopt/gpdbwrappers.cpp
@@ -2845,7 +2845,7 @@ gpdb::GetConstraintRelationOids(Oid constraint_oid, Oid *conrelid,
 	GP_WRAP_END;
 }
 
-Bitmapset *
+List *
 gpdb::GetConstraintRelationColumns(Oid constraint_oid)
 {
 	GP_WRAP_START;
@@ -2853,28 +2853,7 @@ gpdb::GetConstraintRelationColumns(Oid constraint_oid)
 		return get_constraint_relation_columns(constraint_oid);
 	}
 	GP_WRAP_END;
-	return NULL;
-}
-
-void
-gpdb::BmsFree(Bitmapset *a)
-{
-	GP_WRAP_START;
-	{
-		bms_free(a);
-	}
-	GP_WRAP_END;
-}
-
-bool
-gpdb::BmsIsSubset(const Bitmapset *a, const Bitmapset *b)
-{
-	GP_WRAP_START;
-	{
-		return bms_is_subset(a, b);
-	}
-	GP_WRAP_END;
-	return false;
+	return NIL;
 }
 
 // EOF
diff --git a/src/backend/gpopt/translate/CQueryMutators.cpp b/src/backend/gpopt/translate/CQueryMutators.cpp
index 0252500fbe6..4388fca30bc 100644
--- a/src/backend/gpopt/translate/CQueryMutators.cpp
+++ b/src/backend/gpopt/translate/CQueryMutators.cpp
@@ -36,6 +36,22 @@ using namespace gpdxl;
 using namespace gpmd;
 
 
+AttrNumber
+SortGroupClauseContainsPrimaryKey(Query *query, SortGroupClause *sgc, Oid conrelid)
+{
+	Node *expr = gpdb::GetSortGroupClauseExpr(sgc, query->targetList);
+
+	GPOS_ASSERT(IsA(expr, Var));
+
+	Var *var = (Var *) expr;
+
+	RangeTblEntry *rte =
+		(RangeTblEntry *) gpdb::ListNth(query->rtable, var->varno - 1);
+
+	GPOS_ASSERT(NULL != rte);
+
+	return (var->varlevelsup == 0 && rte->relid == conrelid ? var->varattno : InvalidAttrNumber);
+}
 //---------------------------------------------------------------------------
 //	@function:
 //		CQueryMutators::GroupingListContainsPrimaryKey
@@ -48,13 +64,12 @@ using namespace gpmd;
 BOOL
 CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
 											   List *grouping_list,
-											   Bitmapset *conkeys, Oid conrelid)
+											   List *conkeys, Oid conrelid)
 {
-	BOOL result = false;
-	Bitmapset *grouping_set = NULL;
 	ListCell *lgc;
+	ULONG found_col_cnt = 0;
 
-	// Check if conkeys set is a subset of grouping columns set.
+	// search for conkey in grouping list
 	ForEach(lgc, grouping_list)
 	{
 		Node *grpcl = (Node *) lfirst(lgc);
@@ -64,29 +79,14 @@ CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
 			continue;
 		}
 
-		SortGroupClause *sgc = (SortGroupClause *) grpcl;
-		Node *expr = gpdb::GetSortGroupClauseExpr(sgc, query->targetList);
-
-		GPOS_ASSERT(IsA(expr, Var));
-
-		Var *var = (Var *) expr;
-
-		RangeTblEntry *rte =
-			(RangeTblEntry *) gpdb::ListNth(query->rtable, var->varno - 1);
-
-		GPOS_ASSERT(NULL != rte);
-
-		if (0 == var->varlevelsup && rte->relid == conrelid)
+		if (list_member_int(conkeys, SortGroupClauseContainsPrimaryKey(query, (SortGroupClause *) grpcl, conrelid)))
 		{
-			grouping_set = gpdb::BmsAddMember(grouping_set, var->varattno);
+			found_col_cnt++;
 		}
 	}
 
-	result = gpdb::BmsIsSubset(conkeys, grouping_set);
-
-	gpdb::BmsFree(grouping_set);
-
-	return result;
+	// check that all conkeys were found
+	return (gpdb::ListLength(conkeys) == found_col_cnt);
 }
 
 //---------------------------------------------------------------------------
@@ -114,16 +114,10 @@ CQueryMutators::AddMissingGroupClauseMutator(
 			// In case the SortGroupClause is direct element of groupsets list
 			// of GroupingClause, we should treat it in a special way - consider
 			// it like a 1-element list.
-			List *new_list = ListMake1((Node *) gpdb::CopyObject(node));
-			if (GroupingListContainsPrimaryKey(context->m_query, new_list,
-											   context->m_conkeys,
-											   context->m_conrelid))
+			if (1 == list_length(context->m_conkeys) && list_member_int(context->m_conkeys, SortGroupClauseContainsPrimaryKey(context->m_query, (SortGroupClause *) node, context->m_conrelid)))
 			{
-				new_list = gpdb::LAppend(
-					new_list, gpdb::CopyObject((Node *) context->m_gc));
-				return (Node *) new_list;
+				return (Node *) ListMake2((Node *) gpdb::CopyObject(node), gpdb::CopyObject((Node *) context->m_gc));
 			}
-			gpdb::ListFreeDeep(new_list);
 		}
 		return (Node *) gpdb::CopyObject(node);
 	}
@@ -157,7 +151,7 @@ CQueryMutators::AddMissingGroupClauseMutator(
 		ForEach(l, original_list)
 		{
 			Node *n = (Node *) lfirst(l);
-			new_list = gpdb::LAppend(new_list,
+			new_list = list_append_unique(new_list,
 									 AddMissingGroupClauseMutator(n, context));
 		}
 
@@ -475,7 +469,7 @@ CQueryMutators::FixGroupDependentTargets(Query *query)
 			gpdb::ListFreeDeep(old_group_clause);
 			gpdb::GPDBFree(gc);
 		}
-		gpdb::BmsFree(ctx_fix_dependent_targets.m_conkeys);
+		gpdb::ListFree(ctx_fix_dependent_targets.m_conkeys);
 	}
 	// remove constraintDeps
 	gpdb::ListFree(query->constraintDeps);
diff --git a/src/include/catalog/pg_constraint.h b/src/include/catalog/pg_constraint.h
index 3a9304116eb..bfcdcad8fb5 100644
--- a/src/include/catalog/pg_constraint.h
+++ b/src/include/catalog/pg_constraint.h
@@ -264,7 +264,7 @@ extern void ConstraintSetParentConstraint(Oid childConstrId,
 extern bool ConstraintGetPrimaryKeyOf(Oid relid, AttrNumber attno, 
 					Oid *pkrelid, AttrNumber *pkattno);
 extern void get_constraint_relation_oids(Oid constraint_oid, Oid *conrelid, Oid *confrelid);
-extern Bitmapset *get_constraint_relation_columns(Oid constraint_oid);
+extern List *get_constraint_relation_columns(Oid constraint_oid);
 extern Oid	get_relation_constraint_oid(Oid relid, const char *conname, bool missing_ok);
 extern Oid	get_domain_constraint_oid(Oid typid, const char *conname, bool missing_ok);
 extern Oid	get_relation_idx_constraint_oid(Oid relationId, Oid indexId);
diff --git a/src/include/gpopt/gpdbwrappers.h b/src/include/gpopt/gpdbwrappers.h
index 355c852f8f1..d1b582f6854 100644
--- a/src/include/gpopt/gpdbwrappers.h
+++ b/src/include/gpopt/gpdbwrappers.h
@@ -740,13 +740,7 @@ void GetConstraintRelationOids(Oid constraint_oid, Oid *conrelid,
 							   Oid *confrelid);
 
 // find the columns of the relations to which a constraint refers
-Bitmapset *GetConstraintRelationColumns(Oid constraint_oid);
-
-// free a Bitmapset
-void BmsFree(Bitmapset *a);
-
-// check if A is a subset of B
-bool BmsIsSubset(const Bitmapset *a, const Bitmapset *b);
+List *GetConstraintRelationColumns(Oid constraint_oid);
 
 }  //namespace gpdb
 
diff --git a/src/include/gpopt/translate/CQueryMutators.h b/src/include/gpopt/translate/CQueryMutators.h
index a6b832c961f..09e017b2228 100644
--- a/src/include/gpopt/translate/CQueryMutators.h
+++ b/src/include/gpopt/translate/CQueryMutators.h
@@ -188,8 +188,8 @@ class CQueryMutators
 		Query *m_query;
 		// primary key relation id
 		Oid m_conrelid;
-		// set of columns that define primary key in relation with m_conrelid
-		Bitmapset *m_conkeys;
+		// list of columns that define primary key in relation with m_conrelid
+		List *m_conkeys;
 		// the current query level
 		ULONG m_current_query_level;
 		// new SortGroupClause to be added into groupClause
@@ -205,7 +205,7 @@ class CQueryMutators
 		SContextFixGroupDependentTargets(Query *query)
 			: m_query(query),
 			  m_conrelid(0),
-			  m_conkeys(NULL),
+			  m_conkeys(0),
 			  m_current_query_level(0),
 			  m_gc(NULL),
 			  m_parent_is_grouping_clause(false),
@@ -217,8 +217,7 @@ class CQueryMutators
 private:
 	static BOOL GroupingListContainsPrimaryKey(Query *query,
 											   List *grouping_list,
-											   Bitmapset *conkeys,
-											   Oid conrelid);
+											   List *conkeys, Oid conrelid);
 
 	static BOOL GetGroupUniqueTleReferencesWalker(
 		Node *node, SContexGetGroupUniqueTleWalker *context);
```

Reply:
> Yes, I also think it would be good to add to the tests.

done

Reply:
> > Yes, I also think it would be good to add to the tests.
> 
> done

last test passes with [proposal](https://github.com/arenadata/gpdb/pull/746#discussion_r1526162392)